File: Ch06/06_01/rtb.py
def second_best(prices):
    """Return second best price.

    >>> second_best([2, 1, 3])
    2
    """
    if not prices:
        raise ValueError('second_best on empty prices')

    first = second = prices[0]
    for price in prices[1:]:
        if price > first:
            first, second = price, first
        elif price > second:
            second = price
    return second

File: Ch06/06_01/rtb.go
package rtb

import "fmt"

func SecondBest(prices []int) (int, error) {
	if len(prices) == 0 {
		return 0, fmt.Errorf("SecondBest of empty prices")
	}

	first, second := prices[0], prices[0]
	for _, p := range prices[1:] {
		switch {
		case p > first:
			first, second = p, first
		case p > second:
			second = p
		}
	}

	return second, nil
}
